java/vector/src/main/codegen/templates/UnionVector.java

```diff
@@ -54,14 +65,18 @@
  */
 public class UnionVector implements FieldVector {
 
-  private String name;
-  private BufferAllocator allocator;
-  private Accessor accessor = new Accessor();
-  private Mutator mutator = new Mutator();
+  private static final int MAX_TYPE_ID = 128;
+  private static final int MAX_MINOR_TYPE = 128;
+
+  private final String name;
+  private final BufferAllocator allocator;
+  private final Accessor accessor;
+  private final Mutator mutator;
   int valueCount;
 
-  MapVector internalMap;
-  UInt1Vector typeVector;
+  final BitVector bits;
```

Comment:
Is this called `bits` in other vectors? If so it might be worth renaming them to be more clear across the board.

We should also discuss whether a Union needs to be nullable. I think it probably does for consistency cc @jacques-n

Reply:
`bits` is used in: https://github.com/apache/arrow/blob/master/java/vector/src/main/codegen/templates/NullableValueVectors.java#L57
I can rename it, but is probably cleaner in a separate PR?

The existing UnionVector supports nullable interface already, I keep it unchanged. There are other restriction in the existing UnionVector impl - the child vector must be nullable vectors.

Reply:
Gotcha. I presume in existing Java use (e.g. in Dremio) that their unions are non-nullable, and nullability is addressed in the children. In general I think we need the ability to be nullable at the parent union level

Reply:
Yeah, actually the existing UnionVector are already nullable on the parent level, but it uses type[i] == 0 to check null. 

It seems sub-optimal to keep both validity maps in parent vector and children vectors, but the existing UnionVector java interface already returns nullable children vectors so changing it would break compatibility.

Reply:
@jacques-n @siddharthteotia @StevenMPhillips we're going to need some eyes on these issues -- we should try to make the Union implementation consistent with the Arrow format spec without breaking existing downstream applications (or their binary format). I think that's possible but we will have to be careful